### ISSUE TITLE
Why is .JSON recommended over static code for protobufjs?

### ISSUE THREAD

**Issue by hidesminimally** [protobufjs/protobuf.js#1583]
https://github.com/protobufjsprotobuf.js/tree/master/cli#reflection-vs-static-code

> For production environments it is recommended to bundle all your .proto files to a single .json file, which minimizes the number of network requests and avoids any parser overhead (hint: works with just the light library):

However, I don't understand why this is recommended:
1. The number of network requests would be the same for JSON and compiled source code. In other words, both would just have one request if the user rendered the .proto files into one .js/.json file. 
2. To me, the 'minimal' library suggests a smaller size than 'light' library 

And, as a smaller question, the "tradeoff" of no reflection for stati code seems wrong to me - I've thought of using reflection as a code smell. Am curious why this is listed as a trade off.

**Comment by alexander-fenster** [comment]
I think the question here is not about JSON vs static code, but about JSON vs parsing `*.proto` files in the runtime. In other words, it's `protobuf.Root.fromJSON(require('protos.json'))` vs `protobuf.loadSync('input.proto')`, and JSON is preferred mostly because it's a single file read operation, JSON can be easily bundled, and it does not require `fs` to load.

**Comment by ben-xD** [comment]
I disagree @alexander-fenster , the question was the difference between **JSON** and **static code** (generated javascript files). Parsing proto files is out of the question.

I have the same question at @hidesminimally

**Comment by willstott101** [comment]
The linked section of the readme states that there's a point where generated code for very large protobuf libraries will be larger than the difference in size of the minimal and light libraries.

**Comment by ben-xD** [comment]
From the readme then: 

> While using .proto files directly requires the full library respectively pure reflection/JSON the light library, pretty much all code but the relatively short descriptors is shared.
> 
> Static code, on the other hand, requires just the minimal library, but generates additional source code without any reflection features. This also implies that there is a break-even point where statically generated code becomes larger than descriptor-based code once the amount of code generated exceeds the size of the full respectively light library

it is just me or the first sentence doesn't make sense?

Here's what I gather from the 2nd sentence:

**Static code:** starts off small when you don't have many protobufs, as more protobufs are defined, the static code generated grows linearly. It only needs a minimal library.

JSON: Same as static code (starts off small when you don't have many protobufs, as more protobufs are defined, the static code generated grows linearly.) **but** grows at a slower rate, because there is no boiler plate. It's probably a more dense file.

Here's my diagram 😅
![CleanShot 2021-08-17 at 20 28 34](https://user-images.githubusercontent.com/24711048/129788367-cfec15cc-d228-4e3d-b596-8086ec0b0bab.png)

**Comment by ideasculptor** [comment]
reflection isn't a code smell.  It is just slower than non-reflective code.  But there are times when you MUST use reflection because you simply don't know what you have and you must resolve it dynamically. Classic case would be serializing and deserializing an Any protobuf.  You have a string which identifies the type of a proto, and a bunch of binary data in a buffer.  How do you get an instance of that type so you can deserialize into it in the absence of any reflective information?  Protobuf implementations in other languages give you a registry from which you can resolve the protobuf descriptor, and from that you can create a new instance.  Then you can call a generic protobuf unmarshall function which will deserialize the buffer into the instance. There's nothing like that when using protobufjs and, if you use the static module, you don't even have access to the original name of the protobuf type for comparing with the url in the Any proto.  Even with the minimal reflective info you get with json or the full library, you still cannot resolve the fully qualified name including the namepace, nor is there anything which would map to the URL used to serialize an Any proto. You are left with just building some kind of static map of your own, which will resolve to the correct constructor.  There is no way to build such a map reflectively that i can find.

**Comment by dcodeIO** [comment]
Closing this issue because it has been inactive for more than two years and the project has changed substantially.

If this is still relevant, please open a fresh issue with current reproduction steps, version details, and any updated context.

Please also consider that maintaining protobuf.js takes time and effort. If the project is valuable to you or your organization, sponsorship is appreciated.

Thanks!
